kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/GwtComponentServiceImpl.java

```diff
@@ -244,12 +250,20 @@ public GwtConfigComponent findWireComponentConfigurationFromPid(GwtXSRFToken xsr
                 if (conf != null) {
                     conf.getConfigurationProperties().put(ConfigurationAdmin.SERVICE_FACTORYPID, factoryPid);
                 }
-                if ((conf != null) && (conf.getDefinition() == null)) {
+                if (conf != null && conf.getDefinition() == null) {
                     String temporaryName = String.valueOf(System.nanoTime());
                     cs.createFactoryConfiguration(factoryPid, temporaryName, extraProps, false);
                     try {
-                        // wait for the services to be up
-                        TimeUnit.MILLISECONDS.sleep(500);
+
```

Comment:
The initial idea was to have the worker thread sleep for 500 ms until all the temporary factory services are up and running. So, the execution gets halted here. In case of a Service Tracker, it spawns a new thread that waits for the service but the worker thread in this method will still continue and hence, sometimes it can work and sometimes it cannot.

That means, the asynchronous service call can return before the service tracker actually finds a matching service after waiting for the defined 60000 ms. This is because the service tracker spawns a new thread that waits for 60000 ms before it returns a valid service but asynchronous service call can return anytime.

One idea is to use cyclic barrier or countdown latch to check if the service is up and the barrier is closed or latch is counted down to zero before the worker thread continues to work.